src/envs/blocks.py

```diff
@@ -0,0 +1,439 @@
+        next_state = state.copy()
+        # Can only stack if fingers are closed
+        if state.get(self._robot, "fingers") >= self.open_fingers:
+            return next_state
+        # Check that both blocks exist
+        block = self._get_held_block(state)
+        assert block is not None
+        x, y, z, fingers = action.arr
+        other_block = self._get_highest_block_below(state, x, y, z)
+        if other_block is None:  # no block to stack onto
+            return next_state
+        # Can't stack onto yourself!
+        if block == other_block:
+            return next_state
+        # Need block we're stacking onto to be clear
+        if state.get(other_block, "clear") < self.clear_tol:
+            return next_state
+        # Execute stack by snapping into place
+        cur_x = state.get(other_block, "pose_x")
+        cur_y = state.get(other_block, "pose_y")
+        cur_z = state.get(other_block, "pose_z")
+        next_state.set(block, "pose_x", cur_x)
+        next_state.set(block, "pose_y", cur_y)
+        next_state.set(block, "pose_z", cur_z+CFG.blocks_block_size)
+        next_state.set(block, "held", 0.0)
+        next_state.set(block, "clear", 1.0)
+        next_state.set(other_block, "clear", 0.0)
+        next_state.set(self._robot, "fingers", fingers)
+        return next_state
+
+    def get_train_tasks(self) -> List[Task]:
+        return self._get_tasks(num_tasks=CFG.num_train_tasks,
+                               possible_num_blocks=CFG.blocks_num_blocks_train,
+                               rng=self._train_rng)
+
+    def get_test_tasks(self) -> List[Task]:
+        return self._get_tasks(num_tasks=CFG.num_test_tasks,
+                               possible_num_blocks=CFG.blocks_num_blocks_test,
+                               rng=self._test_rng)
+
+    @property
+    def predicates(self) -> Set[Predicate]:
+        return {self._On, self._OnTable, self._GripperOpen, self._Holding,
+                self._Clear}
+
+    @property
+    def types(self) -> Set[Type]:
+        return {self._block_type, self._robot_type}
+
+    @property
+    def options(self) -> Set[ParameterizedOption]:
+        return {self._Pick, self._Stack, self._PutOnTable}
+
+    @property
+    def action_space(self) -> Box:
+        # dimensions: [x, y, z, fingers]
+        return Box(-100.0, 100.0, (4,))
+
+    def render(self, state: State, task: Task,
+               action: Optional[Action] = None) -> Image:
+        raise NotImplementedError
+
+    def _get_tasks(self, num_tasks: int, possible_num_blocks: List[int],
+                   rng: np.random.Generator) -> List[Task]:
+        tasks = []
+        for _ in range(num_tasks):
+            num_blocks = rng.choice(possible_num_blocks)
+            piles = self._sample_initial_piles(num_blocks, rng)
+            init_state = self._sample_state_from_piles(piles, rng)
+            atoms = utils.abstract(init_state, self.predicates)
```

Comment:
i'm thinking about whether this will be an issue when we turn off predicates for learning. probably not, because we're always going to assume that we have the predicates for defining the goals, and that's all that we need here. right?